### PR TITLE
Restore contract size checks

### PR DESCRIPTION
**Pull request by ascjones**
Previously the `ink-waterfall` step reported changes in contract sizes for a PR compared to `master`. This was useful for preventing regressions and also for PRs which are designed to improve the contract sizes.

This PR restores this check, this time integrated as part of the `ink` CI itself. 

I have split the check into separate script files that can be run locally (apart from `contract_sizes_submit.sh`) for easy testing and to make it easy to migrate to GHA later.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑\n     These are the results when building the `integration-tests/*` contracts from this branch with `cargo-contract 4.0.0-alpha-acc02b1` and comparing them to ink! `master`: \n\n    | Contract                                                                  | Optimized Size (Upstream) | Optimized Size (PR) | Diff (kB) | Diff (%)  |
| ------------------------------------------------------------------------- | ------------------------- | ------------------- | --------- | --------- |
| integration-tests/basic-contract-caller                                   | 2.998                     | 2.998               | 0         | 0         |
| integration-tests/basic-contract-caller/other-contract                    | 1.331                     | 1.331               | 0         | 0         |
| integration-tests/call-builder-return-value                               | 8.734                     | 8.734               | 0         | 0         |
| integration-tests/call-runtime                                            | 1.77                      | 1.77                | 0         | 0         |
| integration-tests/conditional-compilation                                 | 1.204                     | 1.204               | 0         | 0         |
| integration-tests/contract-terminate                                      | 1.087                     | 1.087               | 0         | 0         |
| integration-tests/contract-transfer                                       | 1.446                     | 1.446               | 0         | 0         |
| integration-tests/custom-allocator                                        | 7.39                      | 7.39                | 0         | 0         |
| integration-tests/dns                                                     | 7.313                     | 7.313               | 0         | 0         |
| integration-tests/e2e-call-runtime                                        | 1.057                     | 1.057               | 0         | 0         |
| integration-tests/e2e-runtime-only-backend                                | 1.633                     | 1.633               | 0         | 0         |
| integration-tests/erc1155                                                 | 14.113                    | 14.107              | -0.006    | -0.042514 |
| integration-tests/erc20                                                   | 6.572                     | 6.572               | 0         | 0         |
| integration-tests/erc721                                                  | 9.547                     | 9.547               | 0         | 0         |
| integration-tests/events                                                  | 4.816                     | 4.816               | 0         | 0         |
| integration-tests/flipper                                                 | 1.387                     | 1.387               | 0         | 0         |
| integration-tests/incrementer                                             | 1.217                     | 1.217               | 0         | 0         |
| integration-tests/lang-err-integration-tests/call-builder-delegate        | 2.321                     | 2.321               | 0         | 0         |
| integration-tests/lang-err-integration-tests/call-builder                 | 4.875                     | 4.875               | 0         | 0         |
| integration-tests/lang-err-integration-tests/constructors-return-value    | 1.772                     | 1.772               | 0         | 0         |
| integration-tests/lang-err-integration-tests/contract-ref                 | 4.363                     | 4.363               | 0         | 0         |
| integration-tests/lang-err-integration-tests/integration-flipper          | 1.565                     | 1.565               | 0         | 0         |
| integration-tests/mapping-integration-tests                               | 2.975                     | 2.975               | 0         | 0         |
| integration-tests/mother                                                  | 9.499                     | 9.499               | 0         | 0         |
| integration-tests/multi-contract-caller                                   | 5.977                     | 5.977               | 0         | 0         |
| integration-tests/multi-contract-caller/accumulator                       | 1.09                      | 1.09                | 0         | 0         |
| integration-tests/multi-contract-caller/adder                             | 1.672                     | 1.672               | 0         | 0         |
| integration-tests/multi-contract-caller/subber                            | 1.693                     | 1.693               | 0         | 0         |
| integration-tests/multisig                                                | 21.406                    | 21.406              | 0         | 0         |
| integration-tests/payment-channel                                         | 5.544                     | 5.544               | 0         | 0         |
| integration-tests/sr25519-verification                                    | 0.865                     | 0.865               | 0         | 0         |
| integration-tests/static-buffer                                           | 1.411                     | 1.411               | 0         | 0         |
| integration-tests/trait-dyn-cross-contract-calls                          | 2.491                     | 2.491               | 0         | 0         |
| integration-tests/trait-dyn-cross-contract-calls/contracts/incrementer    | 1.3                       | 1.3                 | 0         | 0         |
| integration-tests/trait-erc20                                             | 6.956                     | 6.956               | 0         | 0         |
| integration-tests/trait-flipper                                           | 1.204                     | 1.204               | 0         | 0         |
| integration-tests/trait-incrementer                                       | 1.365                     | 1.365               | 0         | 0         |
| integration-tests/upgradeable-contracts/delegator                         | 2.902                     | 2.902               | 0         | 0         |
| integration-tests/upgradeable-contracts/delegator/delegatee               | 1.368                     | 1.368               | 0         | 0         |
| integration-tests/upgradeable-contracts/set-code-hash                     | 1.46                      | 1.46                | 0         | 0         |
| integration-tests/upgradeable-contracts/set-code-hash/updated-incrementer | 1.439                     | 1.439               | 0         | 0         |
| integration-tests/wildcard-selector                                       | 2.619                     | 2.619               | 0         | 0         |\n\n[Link to the run](https://gitlab.parity.io/parity/ink/-/pipelines/401166) | Last update: Mon Oct 16 15:43:35 CEST 2023

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑
⚠️ **The ink! `master` is ahead of your branch, this might skew the comparison data below.**\n\nThese are the results when building the `integration-tests/*` contracts from this branch with `cargo-contract 4.0.0-alpha-acc02b1` and comparing them to ink! `master`:

| Contract                                                                  | Optimized Size (Upstream) | Optimized Size (PR) | Diff (kB) | Diff (%)  |
| ------------------------------------------------------------------------- | ------------------------- | ------------------- | --------- | --------- |
| integration-tests/basic-contract-caller                                   | 2.998                     | 2.998               | 0         | 0         |
| integration-tests/basic-contract-caller/other-contract                    | 1.331                     | 1.331               | 0         | 0         |
| integration-tests/call-builder-return-value                               | 8.734                     | 8.734               | 0         | 0         |
| integration-tests/call-runtime                                            | 1.77                      | 1.77                | 0         | 0         |
| integration-tests/conditional-compilation                                 | 1.204                     | 1.204               | 0         | 0         |
| integration-tests/contract-terminate                                      | 1.087                     | 1.087               | 0         | 0         |
| integration-tests/contract-transfer                                       | 1.446                     | 1.446               | 0         | 0         |
| integration-tests/custom-allocator                                        | 7.39                      | 7.39                | 0         | 0         |
| integration-tests/dns                                                     | 7.313                     | 7.313               | 0         | 0         |
| integration-tests/e2e-call-runtime                                        | 1.057                     | 1.057               | 0         | 0         |
| integration-tests/e2e-runtime-only-backend                                | 1.633                     | 1.633               | 0         | 0         |
| integration-tests/erc1155                                                 | 14.113                    | 14.107              | -0.006    | -0.042514 |
| integration-tests/erc20                                                   | 6.572                     | 6.572               | 0         | 0         |
| integration-tests/erc721                                                  | 9.547                     | 9.547               | 0         | 0         |
| integration-tests/events                                                  | 4.816                     | 4.816               | 0         | 0         |
| integration-tests/flipper                                                 | 1.387                     | 1.387               | 0         | 0         |
| integration-tests/incrementer                                             | 1.217                     | 1.217               | 0         | 0         |
| integration-tests/lang-err-integration-tests/call-builder-delegate        | 2.321                     | 2.321               | 0         | 0         |
| integration-tests/lang-err-integration-tests/call-builder                 | 4.875                     | 4.875               | 0         | 0         |
| integration-tests/lang-err-integration-tests/constructors-return-value    | 1.772                     | 1.772               | 0         | 0         |
| integration-tests/lang-err-integration-tests/contract-ref                 | 4.363                     | 4.363               | 0         | 0         |
| integration-tests/lang-err-integration-tests/integration-flipper          | 1.565                     | 1.565               | 0         | 0         |
| integration-tests/mapping-integration-tests                               | 2.975                     | 2.975               | 0         | 0         |
| integration-tests/mother                                                  | 9.499                     | 9.499               | 0         | 0         |
| integration-tests/multi-contract-caller                                   | 5.977                     | 5.977               | 0         | 0         |
| integration-tests/multi-contract-caller/accumulator                       | 1.09                      | 1.09                | 0         | 0         |
| integration-tests/multi-contract-caller/adder                             | 1.672                     | 1.672               | 0         | 0         |
| integration-tests/multi-contract-caller/subber                            | 1.693                     | 1.693               | 0         | 0         |
| integration-tests/multisig                                                | 21.406                    | 21.406              | 0         | 0         |
| integration-tests/payment-channel                                         | 5.544                     | 5.544               | 0         | 0         |
| integration-tests/sr25519-verification                                    | 0.865                     | 0.865               | 0         | 0         |
| integration-tests/static-buffer                                           | 1.411                     | 1.411               | 0         | 0         |
| integration-tests/trait-dyn-cross-contract-calls                          | 2.491                     | 2.491               | 0         | 0         |
| integration-tests/trait-dyn-cross-contract-calls/contracts/incrementer    | 1.3                       | 1.3                 | 0         | 0         |
| integration-tests/trait-erc20                                             | 6.956                     | 6.956               | 0         | 0         |
| integration-tests/trait-flipper                                           | 1.204                     | 1.204               | 0         | 0         |
| integration-tests/trait-incrementer                                       | 1.365                     | 1.365               | 0         | 0         |
| integration-tests/upgradeable-contracts/delegator                         | 2.902                     | 2.902               | 0         | 0         |
| integration-tests/upgradeable-contracts/delegator/delegatee               | 1.368                     | 1.368               | 0         | 0         |
| integration-tests/upgradeable-contracts/set-code-hash                     | 1.46                      | 1.46                | 0         | 0         |
| integration-tests/upgradeable-contracts/set-code-hash/updated-incrementer | 1.439                     | 1.439               | 0         | 0         |
| integration-tests/wildcard-selector                                       | 2.619                     | 2.619               | 0         | 0         |

[Link to the run](https://gitlab.parity.io/parity/ink/-/pipelines/401214) | Last update: Mon Oct 16 17:15:44 CEST 2023

File: scripts/markdown-table.sh [hunk 1]
```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env bash
+
+# Script copied from: https://josh.fail/2022/pure-bash-markdown-table-generator/
```
Comment (ascjones) [inline scripts/markdown-table.sh:3]: Note this whole script to generate a markdown table copied from https://josh.fail/2022/pure-bash-markdown-table-generator/

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -412,7 +412,51 @@ examples-docs:
...
-#### stage:                        ink-waterfall
+#### stage:                        measurements
+
+contract-sizes:
+  stage:                           measurements
+  <<:                              *docker-env
+  rules:
+    - if: $CI_PIPELINE_SOURCE == "push" && $CI_COMMIT_REF_NAME != "master"
+  variables:
+    # Fix for linking of `linkme` for `cargo contract build`: https://github.com/dtolnay/linkme/issues/49
```
Comment (cmichi) [inline .gitlab-ci.yml:423]: Why is `linkme` involved in our CI in the first place?

Comment (ascjones) [reply .gitlab-ci.yml:423]: It's to do with the events 2.0 metadata, there is a fix now so maybe we don't need that at all anymore, see https://github.com/paritytech/ink/pull/1921 and https://github.com/paritytech/cargo-contract/pull/1344

**Pull request merged**